mtags/src/main/scala-2/scala/meta/internal/pc/SemanticTokenProvider.scala

```diff
@@ -0,0 +1,425 @@
+      case _: Token.KwFalse   => getTypeId(SemanticTokenTypes.Keyword)
+
+      // extends Symbolic keywords
+      case _: Token.Hash => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.Viewbound => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.LeftArrow => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.Subtype => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.RightArrow => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.Supertype => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.At => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.Underscore => getTypeId(SemanticTokenTypes.Variable)
+      case _: Token.TypeLambdaArrow => getTypeId(SemanticTokenTypes.Operator)
+      case _: Token.ContextArrow => getTypeId(SemanticTokenTypes.Operator)
+
+      // Constant
+      case _: Token.Constant.Int | _: Token.Constant.Long |
+          _: Token.Constant.Float | _: Token.Constant.Double =>
+        getTypeId(SemanticTokenTypes.Number)
+      case _: Token.Constant.String | _: Token.Constant.Char =>
+        getTypeId(SemanticTokenTypes.String)
+      case _: Token.Constant.Symbol => getTypeId(SemanticTokenTypes.Property)
+
+      // Comment
+      case _: Token.Comment => getTypeId(SemanticTokenTypes.Comment)
+
+      // Interpolation 
+      case _: Token.Interpolation.Id
+          |_: Token.Interpolation.SpliceStart
+             => getTypeId(SemanticTokenTypes.Keyword)
+      case _: Token.Interpolation.Start
+          |_: Token.Interpolation.Part
+          |_: Token.Interpolation.SpliceEnd
+          |_: Token.Interpolation.End
+             => getTypeId(SemanticTokenTypes.String)  // $ symbol
+
+      // Default
+      case _ => -1
+    }
+
+  }
+  def pickFromTraversed(tk: scala.meta.tokens.Token): Option[NodeInfo] = {
+
+    val buffer = ListBuffer.empty[NodeInfo]
+    for (node <- nodes) {
+      if (
+        node.pos.start == tk.pos.start &&
+        node.pos.end == tk.pos.end
+      ) buffer.++=(List(node))
+    }
+
+    buffer.toList.headOption
+  }
+
+  case class NodeInfo(
+      sym: Option[Symbol],
+      pos: scala.reflect.api.Position
+  )
+  object NodeInfo {
+    def apply(tree: Tree, pos: scala.reflect.api.Position): NodeInfo =
+      new NodeInfo(Some(tree.symbol), pos)
+
+    def apply(sym: Symbol, pos: scala.reflect.api.Position): NodeInfo =
+      new NodeInfo(Some(sym), pos)
+  }
+
+  /**
+   * was written in reference to PcDocumentHighlightProvider.
+   */
+  import cp._
+  object traverser {
+
+    /**
+     * gathers all nodes inside given tree.
+     * The nodes have symbol.
+     */
+    def traverse(
+        nodes: List[NodeInfo],
+        tree: cp.Tree
+    ): List[NodeInfo] = {
+
+      tree match {
+        /**
+         * All indentifiers such as:
+         * val a = <<b>>
+         */
+        case ident: cp.Ident if ident.pos.isRange =>
+          nodes :+ NodeInfo(ident, ident.pos)
```

Comment:
```suggestion
          val symbol =
            if (ident.symbol == NoSymbol) 
              if(ident.tpe != null)
              ident.tpe.typeSymbol
              else {
                val context = doLocateContext(ident.pos)
                context.lookupSymbol(ident.name, _ => true) match {
                   case LookupSucceeded(_, symbol) => symbol
                   case _ => NoSymbol
                }
              }
            else ident.symbol
          nodes :+ NodeInfo(symbol, ident.pos)
```
this should help with StructuralTypes case.